### ISSUE TITLE
Pre-build and Serialize Plaquettes

### ISSUE THREAD

**Issue by inmzhang** [tqec/tqec#403]
Currently, we build plaquettes on the fly during the compilation of the `BlockGraph`. The code used to build the plaquettes is located in `src/tqec/plaquette/library`. Since there are many variants of similar plaquettes that use the same piece of builder code, the number of arguments and the code size keep growing, making it hard to follow and manage the complexity.

One alternative approach is to pre-build the plaquettes, serialize them, and store them in the library. Since we’ve decided to use the name as the unique identifier for the plaquette and hash it during detector computation, it would be natural to maintain a database-like library for the plaquettes. This would also allow us to remove the complex builder code or move it to independent packages. We could then select which plaquette to use during compilation by querying it by name.

One major advantage of this approach, in my view, is that we can build the plaquettes with Crumble, which will be much easier for small local plaquette circuits. This would also make it more accessible for newcomers to contribute a plaquette.

**Comment by Gistbatch** [comment]
I think we originally wanted to use the front end we were building to generate this library.
For #273 there was already a discussion about implementing a database, so I think this is generally a good idea.

I still think we should have builder code "somewhere" for the moment.
We could also use the database to cache other things, although this would introduce a couple of new dependencies I think. Alternatively, we could implement this as a standalone thing, but this might be overkill.

**Comment by inmzhang** [comment]
> For https://github.com/tqec/tqec/issues/273 there was already a discussion about implementing a database, so I think this is generally a good idea.

By "database-like" I intended to mean some json(or some other formats) files that includes human-readable qubits and `stim` circuits. And the files can be indexed by their name, which should be the same as the unique identifier of the plaquettes. I think a SQL database for the plaquettes can be overkill until the set of plaquettes grows too large.

**Comment by Gistbatch** [comment]
My view on this was the following:

1. Plaquettes should not be created inside `tqec` except for a restricted set of defaults.
Instead, there should be other ways to import this, either through `crumble` or our frontend (if this ever gets done)
2. In my opinion there should be a database, which could be as simple as an object store, where we can look up existing plaquettes in an easy way. In the best case, this would be centralized, such that all plaquettes created by users can be reused. Hence, why I wanted to make this independent.
3. We should still have a way to programmatically generate plaquettes for providing default plaquettes. I'd prefer this to be decoupled though, so regular users won't need to deal with this.

With your long-term vision, this would add a further point:
- `tqec.plaquette.library` contains default plaquettes and their creation code
- `tqec.plaquette.database` contains a collection of independent plaquettes which can be queried/reused in `Template`

**Comment by inmzhang** [comment]
> The above names are only examples and might not be descriptive enough, but I think having such a representation could solve https://github.com/tqec/tqec/issues/386 and help us visualizing collection of plaquettes.

One problem I have with using `function` strings to index the plaquettes in the template is that the function of a plaquette placeholder in the template can change. For example, for a `QubitTemplate`, an index can represent either a `XXXX` or `ZZZZ` plaquette depending on it's within a `ZXZ` or `XZX` cube(the boundary types have changed). Did I miss anything?

**Comment by nelimee** [comment]
> > The above names are only examples and might not be descriptive enough, but I think having such a representation could solve #386 and help us visualizing collection of plaquettes.
> 
> One problem I have with using `function` strings to index the plaquettes in the template is that the function of a plaquette placeholder in the template can change. For example, for a `QubitTemplate`, an index can represent either a `XXXX` or `ZZZZ` plaquette depending on it's within a `ZXZ` or `XZX` cube(the boundary types have changed). Did I miss anything?

This is also one of the main issue I have with the solution I proposed. It can be solved using several methods:

- `function` names do not depends on the basis, but rather use something like "basis_1" and "basis_2" and it is up to the template to adapt that.
- `Template.instantiate` returns an object instance that can be mutated (e.g., changing the basis / boundary) before actually returning the 2D array of `function` entries.
- `Template.instantiate` take more parameters to account for those cases. 

The most important thing in this idea is to agree on the naming and how we will handle such subtleties, but I think the advantages (not having to hand-write a `Plaquettes` instance for each `Template` sub-class) are still worth the hassle.



> My view on this was the following:
> 
>     1. Plaquettes should not be created inside `tqec` except for a restricted set of defaults.
>        Instead, there should be other ways to import this, either through `crumble` or our frontend (if this ever gets done)
> 
>     2. In my opinion there should be a database, which could be as simple as an object store, where we can look up existing plaquettes in an easy way. In the best case, this would be centralized, such that all plaquettes created by users can be reused. Hence, why I wanted to make this independent.
> 
>     3. We should still have a way to programmatically generate plaquettes for providing default plaquettes. I'd prefer this to be decoupled though, so regular users won't need to deal with this.
> 
> 
> With your long-term vision, this would add a further point:
> 
>     * `tqec.plaquette.library` contains default plaquettes and their creation code
> 
>     * `tqec.plaquette.database` contains a collection of independent plaquettes which can be queried/reused in `Template`

Now that `Plaquette` instances are immutable, I am fine with the idea and I agree with your long-term vision.
